### DISCUSSION TITLE
How to handle IDs with custom generation logic?

### DISCUSSION THREAD

**Discussion by preslavrachev** [sqlc-dev/sqlc#1354] [Q&A]
Hello, 

I am relatively fresh to sqlc, so please forgive me if I am asking something obvious. We are in considerations of replacing Gorm with sqlc for our production applications, however we have identified one tiny issue. Many of our tables do not follow the standard auto-increment ID logic but are custom varchars, based on certain conditions.

In Gorm, we could tackle this easily through the `BeforeCreate` hooks. However, in the way sqlc creates models, we don't seem to have much control over the ID generation logic (other than purely addressing it on the Postgres side).

What would advise me to do? I saw that the configuration allows telling the generator to use predefined custom types for some of the table columns. Is this the way to go? If we create custom `ID` types on our side which are based on `string` or conform to the `Stringer` interface, we should in theory be able to generate custom IDs at the time of querying. Am I right?

**Comment by Jille** [comment]
The best I can think of is adding another file next to the generated files and implement `func (q *Queries) CreateMyObject()` that sets the id and then calls the generated method.

**Comment by skabbes** [comment]
> Many of our tables do not follow the standard auto-increment ID logic but are custom varchars, based on certain conditions.

sqlc doesn't really care if it is auto-incrementing or not, integer, or text or anything.  Since sqlc is "just SQL" - All it cares about it whether the ID is generated by sql or not.

For example: here is using a postgres function to generate an ID

```sql
create table thing (
  id TEXT NOT NULL,
  name TEXT NOT NULL
);

-- name: CreateThing :one
INSERT INTO thing (id, name)
VALUES(gen_random_uuid(), sqlc.arg('name'))
RETURNING *;
```

Or you could also do what @Jille suggested.

```sql
-- name: CreateThing :one
INSERT INTO thing (id, name)
VALUES(sqlc.arg('id'), sqlc.arg('name'))
RETURNING *;
```

And then just wrap the function yourself.

```go
// go boilerplate omitted
func CreateThing(name: string) {
  id := generateId();
  queries.CreateThing(CreateThingParams{ID: id, Name: name})
}

```

I think if you are _just_ using sqlc to create model objects, you will likely think sqlc is lacking a lot of features. The true power is by leveraging SQL itself (in my opinion).
